### ISSUE TITLE
Feature/improvement brainstorming

### ISSUE THREAD

**Issue by ivoanjo** [ivoanjo/gvl-tracing#13]
I've spent some time thinking about potential improvements to gvl-tracing, and decided to record it here. If anyone's interested in tackling them, feel free to ask for details!

* (Easy) Have a block-based version for `GvlTracing.start`, allowing

  ```ruby
  GvlTracing.start("somefile.json") do
    # Code that gets traced goes in here
  end
  ```

* (Medium) Create a wrapper command to run some other Ruby command with gvl-tracing:

  ```bash
  $ bundle exec gvl-tracing ruby some-example.rb
  ```

* (Medium/Hard) Fix thread reuse. Because [Ruby reuses native os threads](https://ivoanjo.me/blog/2022/11/26/ruby-reuses-native-threads/) the thread ids/names in gvl-tracing can get incorrectly reused across different Ruby Threads.

* (Hard) Add support for producing data in a format readable by the [Firefox Profiler](https://profiler.firefox.com/) or [Speedscope](https://www.speedscope.app/). Both provide interesting alternatives to perfetto and seem to be easier to integrate/share data with.

* (Medium) Improved thread naming
  * Issue #9
  * In progress: #12

* (Medium) Add tests! Because I built this gem as an experiment over a few nights and weekends, it was small enough that I could get away without a test suite. I would mega appreciate if someone helped change that :)

**Comment by benoittgt** [comment]
For the test do you have something in mind? I thought about something simple base on examples folder + partial match of the output. More like integration testing.

**Comment by ivoanjo** [comment]
Yes, I was thinking along those lines of having small code blocks to try to exercise the gem with something that produces only a few events, and then parse the output and see if it looks reasonable.

**Comment by benoittgt** [comment]
Also another idea, will be to script the examples generation + json compacting.

**Comment by ivoanjo** [comment]
Updated list of tasks now that #12 and #14 fix some of them :)

**Comment by ivoanjo** [comment]
Added reference to https://github.com/ruby/ruby/pull/8885 since we can use it to solve the thread reuse issue :)

**Comment by jpcamara** [comment]
Hi there 🙋🏻‍♂️! I'll be submitting some PRs to:

- Add some specs
- Attempt Firefox profiler support (it's a little tricky but I do have it loading in the profiler)
- Fix Ruby 3.3 support - because the parent thread can fire the events now, getting the current thread id causes the timeline to show the Main thread as starting multiple times and the other threads never show as starting

Should have at least the specs up in the next day or so.

**Comment by ivoanjo** [comment]
THAT IS AMAZING AND I CAN'T WAIT!

I'll add the caveat that I moved houses the past weekend and I may take a bit to review stuff, but I definitely want to :)

**Comment by jpcamara** [comment]
> THAT IS AMAZING AND I CAN'T WAIT!
> 
> I'll add the caveat that I moved houses the past weekend and I may take a bit to review stuff, but I definitely want to :)

Haha understandable! I'm sure it'll take me a few days to get everything through anyways. Good luck with the post-move tasks 🫡

**Comment by jpcamara** [comment]
I haven't lost sight of the 3.3 updates btw! I'm finishing up some work on MN thread support for Mac in Ruby (https://github.com/ruby/ruby/pull/9178) then I'll open a 3.3 PR here

**Comment by ivoanjo** [comment]
Sounds great :bow:!

**Comment by ivoanjo** [comment]
Updated list of ideas after 1.5.0 release! :)
